### ISSUE TITLE
[CPU][AMX] brg_matmul bf16 kernel leaks NaN/garbage data into unrelated  output rows when K dimension is not a multiple of 32

### ISSUE THREAD

**Issue by LshhsLup** [oneapi-src/oneDNN#5717]
## Summary

When running bf16 matmul on CPU using the AMX-accelerated kernel (`brg_matmul:avx10_1_512_amx`), if the K dimension is **not a multiple of 32**, NaN values present in *unrelated* rows of the input tensor leak into output rows that should be numerically independent and NaN-free.

Mathematically, `output[i, :]` of a matmul should only depend on `input[i, :]`. In this bug, a row of the input that contains **no NaN** still produces a NaN-contaminated output row, apparently due to data leaking in from other rows/batches during AMX tile handling of the K-dimension remainder ("tail") when `K % 32 != 0`.

This was originally discovered via PyTorch's `torch.bmm` CPU bf16 path, and has been independently reproduced with:
- `benchdnn` (confirms the exact same kernel is selected for this shape)
- A minimal, pure oneDNN C++ API program with no PyTorch involvement at all

## Environment

- **oneDNN version:** v3.14.0 (commit `b7bfe8bf3c252d31bcbf93d30b617897a1e447be`)
- **OS:** Ubuntu 22.04.5 LTS (Jammy Jellyfish), kernel `6.2.0-26-generic`
- **CPU:** Intel(R) Xeon(R) Gold 6430 (Sapphire Rapids), 2 sockets × 32 cores  (128 threads total), Family 6 Model 143 Stepping 8
- **Relevant CPU flags:** `amx_bf16`, `amx_tile`, `amx_int8`, `avx512_bf16`,  `avx512_fp16`
- **oneDNN runtime ISA detection:** "Intel AVX10.1 and Intel AMX with bfloat16 and 8-bit integer support"
- **Compiler:** g++ 11.4.0 (Ubuntu 11.4.0-1ubuntu1~22.04.3)
- **CMake:** 3.22.1
- **Runtime:** OpenMP
- **Container context:** Running inside a Docker container. CPU instructions are passed through  directly from the host kernel — no full hardware virtualization/hypervisor is involved, so this should not be related to VM-level AMX tile register state handling. Host CPU is bare-metal Sapphire Rapids Xeon Gold 6430.
- **Note:** The minimal C++ reproduction below is a small single-threaded program; the OpenMP thread count shown in verbose logs reflects the library's default runtime configuration, not necessarily active parallelism during this specific repro.

## Root cause investigation: K-dimension sweep

Using a batched matmul shape of `(batch=16, M=32, K=D, N=64)`, where all rows of the input except row 0 are set to all-NaN, and row 0 is kept
NaN-free, we swept `D` across a range of values and checked whether row 0's *output* contained any NaN (it should never contain NaN, since row 0's input is clean):

| D  | multiple of 32? | Row 0 output polluted with NaN? |
|----|:---:|:---:|
| 16 | — (<32) | No |
| 17 | — (<32) | No |
| 31 | — (<32) | No |
| 32 | yes | No |
| **33** | no | **Yes** |
| **34** | no | **Yes** |
| **40** | no | **Yes** |
| **47** | no | **Yes** |
| **48** | no | **Yes** |
| **49** | no | **Yes** |
| **63** | no | **Yes** |
| 64 | yes | No |
| **65** | no | **Yes** |
| **95** | no | **Yes** |
| 96 | yes | No |
| **97** | no | **Yes** |
| 128 | yes | No |

**The pattern is exact and 100% reproducible:** the bug triggers if and only if `D > 32` and `D` is **not** a multiple of 32. Any `D <= 32`, or any `D` that is an exact multiple of 32, produces correct (NaN-free) output in row 0.

This strongly suggests that:
1. The AMX tile's native bf16 blocking factor along the K dimension is 32.
2. When `K` is a multiple of 32 (or fits in a single tile, `K <= 32`), the blocking is exact and no special handling is needed.
3. When `K > 32` and `K % 32 != 0`, oneDNN must handle a "tail" (remainder) tile smaller than 32 elements — and this tail-handling path appears to read uninitialized or unrelated memory (e.g. adjacent rows/batches containing NaN) instead of correctly zero-padding or isolating the partial tile.

Supporting evidence: the internal blocked memory format selected by oneDNN for this shape is named `aCB16b32c` (see verbose logs below) — the `32c` suffix is consistent with a 32-element blocking factor along one dimension, matching the observed threshold exactly.

## Reproduction 1: PyTorch (where this was first observed)

```python
import torch

N, H, D, K = 32, 16, 33, 64  # D=33 triggers the bug; D=32 does not

def run(dtype: torch.dtype, seed: int = 0) -> torch.Tensor:
    g = torch.Generator().manual_seed(seed)
    q = torch.randn(N, H, D, generator=g)          # (N, H, D)
    wc = torch.rand(H, K, D, generator=g)           # (H, K, D)
    # Make every row except row 0 all-NaN.
    nan_rows = [i for i in range(N) if i != 0]
    q[nan_rows] = torch.nan
    assert not q[0].isnan().any(), "fixture: row 0 must be NaN-free"
    assert not wc.isnan().any(), "fixture: weight must be NaN-free"

    out = torch.bmm(
        q.to(dtype).transpose(1, 0),                # (H, N, D)
        wc.to(dtype).permute(0, 2, 1),               # (H, D, K)
    )                                                # -> (H, N, K)
    return out

bf16_out = run(torch.bfloat16)
fp32_out = run(torch.float32)

print("bf16 row0 has NaN:", bf16_out[:, 0, :].isnan().any().item())  # True  (BUG)
print("fp32 row0 has NaN:", fp32_out[:, 0, :].isnan().any().item())  # False (correct)
```

**PyTorch version:** 2.11.0+cpu

Disabling mkldnn confirms the issue is entirely within the oneDNN backend:

```python
torch.backends.mkldnn.enabled = False
# re-running the above now gives bf16 row0 has NaN: False (correct)
```

PyTorch-side verbose log for this shape:

```
onednn_verbose,v1,primitive,exec,cpu,matmul,brg_matmul:avx10_1_512_amx,undef,src:bf16::blocked:abc::f0 wei:bf16::blocked:acb::f0 dst:bf16::blocked:abc::f0,attr-scratchpad:user,,16x32x33:16x33x64,5.87598
```

## Reproduction 2: benchdnn (confirms same kernel is used, no NaN injected)

```bash
ONEDNN_VERBOSE=1 ./tests/benchdnn/benchdnn --matmul \
  --dt=bf16:bf16:bf16 \
  --stag=abc --wtag=acb --dtag=abc \
  16x32x33:16x33x64
```

Output (trimmed):

```
onednn_verbose,v1,info,oneDNN v3.14.0 (commit b7bfe8bf3c252d31bcbf93d30b617897a1e447be)
onednn_verbose,v1,info,cpu,runtime:OpenMP,nthr:128
onednn_verbose,v1,info,cpu,isa:Intel AVX10.1 and Intel AMX with bfloat16 and 8-bit integer support
onednn_verbose,v1,primitive,exec,cpu,reorder,jit:uni,undef,src:f32::blocked:acb::f0 dst:f32:p:blocked:aCB16b32c::f0,,,16x33x64,14.6279
onednn_verbose,v1,primitive,exec,cpu,matmul,brg_matmul:avx10_1_512_amx,undef,src:bf16::blocked:abc::f0 wei:bf16::blocked:acb::f0 dst:bf16::blocked:abc::f0,,,16x32x33:16x33x64,34.9248
0:PASSED (568 ms) __REPRO: --matmul --dt=bf16:bf16:bf16 --stag=abc --wtag=acb --dtag=abc 16x32x33:16x33x64
```

Notes:
- The kernel selected (`brg_matmul:avx10_1_512_amx`) and memory formats (`abc`/`acb`/`abc`, `aCB16b32c`) exactly match the PyTorch-side verbose log above, confirming both code paths hit the identical implementation.
- benchdnn reports `PASSED` here because its default correctness check only verifies numerical accuracy against a reference implementation using  **clean random inputs** — it does not inject NaN into part of the input to check for cross-row/cross-tile data leakage. This is why the bug is not caught by benchdnn's standard correctness testing, and likely why it has gone unnoticed in existing test coverage.

## Reproduction 3: Minimal pure oneDNN C++ API program (no PyTorch, no benchdnn)

This program directly constructs a bf16 matmul via the oneDNN C++ API, with row 0 set to a clean value (`1.0`) and all other rows set to NaN
(`0x7FC0` bit pattern), then checks whether row 0 of the output is contaminated.

```cpp
#include "oneapi/dnnl/dnnl.hpp"
#include <vector>
#include <cmath>
#include <cstring>
#include <iostream>

using namespace dnnl;

int main() {
    engine eng(engine::kind::cpu, 0);
    stream s(eng);

    const int H = 16, N = 32, D = 33, K = 64;

    memory::dims src_dims = {H, N, D};
    memory::dims wei_dims = {H, D, K};
    memory::dims dst_dims = {H, N, K};

    auto src_md = memory::desc(src_dims, memory::data_type::bf16, memory::format_tag::abc);
    auto wei_md = memory::desc(wei_dims, memory::data_type::bf16, memory::format_tag::acb);
    auto dst_md = memory::desc(dst_dims, memory::data_type::bf16, memory::format_tag::abc);

    auto matmul_pd = matmul::primitive_desc(eng, src_md, wei_md, dst_md);
    auto matmul_prim = matmul(matmul_pd);

    auto src_mem = memory(src_md, eng);
    auto wei_mem = memory(wei_md, eng);
    auto dst_mem = memory(dst_md, eng);

    // Row 0 clean (1.0), all other rows NaN (bf16 NaN bit pattern 0x7FC0).
    std::vector<uint16_t> src_data(H * N * D);
    for (int h = 0; h < H; h++)
        for (int n = 0; n < N; n++)
            for (int d = 0; d < D; d++) {
                if (n == 0)
                    src_data[h*N*D + n*D + d] = 0x3F80; // bf16 of 1.0
                else
                    src_data[h*N*D + n*D + d] = 0x7FC0; // bf16 NaN
            }
    std::memcpy(src_mem.get_data_handle(), src_data.data(), src_data.size()*sizeof(uint16_t));

    std::vector<uint16_t> wei_data(H * D * K, 0x3F80); // all 1.0, NaN-free
    std::memcpy(wei_mem.get_data_handle(), wei_data.data(), wei_data.size()*sizeof(uint16_t));

    matmul_prim.execute(s, {{DNNL_ARG_SRC, src_mem}, {DNNL_ARG_WEIGHTS, wei_mem}, {DNNL_ARG_DST, dst_mem}});
    s.wait();

    uint16_t* dst_data = (uint16_t*)dst_mem.get_data_handle();
    bool row0_has_nan = false;
    for (int h = 0; h < H; h++)
        for (int k = 0; k < K; k++) {
            uint16_t v = dst_data[h*N*K + 0*K + k]; // row 0 of output
            if ((v & 0x7F80) == 0x7F80 && (v & 0x007F) != 0) // bf16 NaN pattern
                row0_has_nan = true;
        }

    std::cout << "Row 0 has NaN: " << (row0_has_nan ? "YES (BUG)" : "NO (correct)") << std::endl;
    return 0;
}
```

Build and run:

```bash
g++ -std=c++17 repro.cpp -I include -I build/include -L build/src -ldnnl -o repro
LD_LIBRARY_PATH=build/src ./repro
```

**Output:**

```
Row 0 has NaN: YES (BUG)
```

This confirms the bug is entirely contained within oneDNN's `brg_matmul:avx10_1_512_amx` kernel — no PyTorch or benchdnn code is
involved in this reproduction.

## Expected behavior

Since `output[i, :]` of a matmul is mathematically independent of any other row `j != i` of the input, row 0's output should never contain NaN when row 0's input is NaN-free — regardless of what values (including NaN) appear in other rows of the batch.

## Actual behavior

When `K % 32 != 0` and `K > 32`, row 0's output is contaminated with NaN, apparently leaked from other rows/batches during AMX tile-based handling of the K-dimension remainder.

## Suggested area to investigate

Given the kernel name (`brg_matmul:avx10_1_512_amx`) and the blocked format name (`aCB16b32c`), this likely points to the K-dimension tail/remainder handling logic in the AMX-based BRGEMM implementation, possibly under `src/cpu/x64/brgemm/` — specifically wherever the implementation decides how to pack/pad a K-tail smaller than the native 32-element AMX tile block for bf16. It's possible the tail tile buffer is not being fully zero-initialized before use, or an incorrect tile size/stride is used when copying the K-remainder into the AMX tile register, causing it to read adjacent memory (other batch/row data) instead of properly isolated, zero-padded data.

I'm happy to help narrow this down further (e.g. additional shape sweeps, thread-count variations, `avx512_core` fallback vs AMX path comparison) if that's useful, and would be interested in attempting a fix with some guidance on the relevant code path, given prior experience with tiled kernel / padding edge cases in accelerator kernel development.

**Comment by vpirogov** [comment]
Looks like we have a bug in NaN handling in `extendable_k` strategy.

+@tprimak, @densamoilov
